core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java

```diff
@@ -285,24 +266,58 @@ private void resetAdapter() {
     searchPresenter.getRecentSearches();
   }
 
-  private ArrayAdapter<String> getDefaultAdapter() {
-    return new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1) {
-      @NonNull
-      @Override
-      public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
-        View row;
+  private DefaultAdapter getDefaultAdapter() {
+    return new DefaultAdapter();
+  }
 
-        if (convertView == null) {
-          row = LayoutInflater.from(parent.getContext())
-            .inflate(android.R.layout.simple_list_item_1, null);
-        } else {
-          row = convertView;
-        }
+  public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
```

Comment:
extract to a separate file

Reply:
Should I move the view holder too? I think yes, the doubt then arises. How should I go with updating searchPresenter.

Reply:
Callbacks probably. Higher order functions if you use kotlin.

Reply:
I don't use Kotlin
I found another way: some thing like this:
`if(context instanceof searchActivity){
((SearchActivity)this).searchPresenter ...
}`

I hope this will work fine.

Reply:
No, grabbing an instance an casting it is pretty messy. You can define an interface and supply an implementation in the constructor.

Consider `View.onSetClickListener`

```
public void setOnClickListener(@Nullable OnClickListener l)
```

You would have used this previously I imagine. What you do here is provide some code to be later executed when a view is clicked

```
    public interface OnClickListener {
        /**
         * Called when a view has been clicked.
         *
         * @param v The view that was clicked.
         */
        void onClick(View v);
    }
```

So we have to supply an implementation of this interface, note _implementation_ because an interface cannot be instantiated.
We can do this with an anonymous class
```
 view.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        //code that gets executed on a click
      }
    });
```
but it is handier if we can use the syntax sugar of a lambda

```
  view.setOnClickListener(v -> {
    //code that gets executed on a click
    });
```

There is nothing special about this, just another object, so you can extract a variable (`ctrl+alt+v`) and now it looks like this
```
    final View.OnClickListener onClickListener = v -> {

    };
    view.setOnClickListener(onClickListener);
```

Objects can of course be passed around so to other objects, composing functionality

```
class ContrivedExample {
  private final OnClickListener listener;
  public ContrivedExample(OnClickListener listener){
    this.listener=listener;
  }

  public bindClickListener(View v){
   v.setOnClickListener(listener) 
  }
}
```

which is very much what you want to do in the default adapter.

I also heavily recommend learning Kotlin, Android is Kotlin First and so is Kiwix.